homeassistant/components/fritzbox/binary_sensor.py

```diff
@@ -68,18 +68,25 @@ async def async_setup_entry(
     hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
 ) -> None:
     """Set up the FRITZ!SmartHome binary sensor from ConfigEntry."""
-    coordinator: FritzboxDataUpdateCoordinator = hass.data[FRITZBOX_DOMAIN][
-        entry.entry_id
-    ][CONF_COORDINATOR]
-
-    async_add_entities(
-        [
-            FritzboxBinarySensor(coordinator, ain, description)
-            for ain, device in coordinator.data.devices.items()
-            for description in BINARY_SENSOR_TYPES
-            if description.suitable(device)
-        ]
-    )
+    coordinator: FritzboxDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id][
+        CONF_COORDINATOR
+    ]
+
+    def _add_entities() -> None:
```

Comment:
Missing @callback marking. Right now its executing in executors.

Reply:
Just have 
devices = set()

locally here instead. Or pass the new devices as an argument to the signal.

Reply:
you mean something like this?

```python
async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
) -> None:
    """Set up the FRITZ!SmartHome binary sensor from ConfigEntry."""
    coordinator: FritzboxDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id][
        CONF_COORDINATOR
    ]
    added_devices: list[str] = hass.data[DOMAIN][entry.entry_id][PLATFORM_DOMAIN]

    def _add_entities() -> None:
        """Add devices."""
        entities: list[FritzboxBinarySensor] = []
        for ain, device in coordinator.data.devices.items():
            if ain in added_devices:
                continue
            added_devices.append(ain)
            for description in BINARY_SENSOR_TYPES:
                if description.suitable(device):
                    entities.append(FritzboxBinarySensor(coordinator, ain, description))

        async_add_entities(entities)

    entry.async_on_unload(coordinator.async_add_listener(_add_entities))

    _add_entities()
```

Reply:
added_devices = [] is perfectly enough. It does not need to be shared between platforms.

Reply:
you never stop learning 🤓 i assumed that i've to "store `added_devices` somewhere" to have it persistent for upcoming calls via a signal call 🙈

Reply:
`@callback` is still missing